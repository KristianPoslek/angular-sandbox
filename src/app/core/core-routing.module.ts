import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';

const routes: Routes = [
    {
        path: '',
        loadChildren: '../pokemon-list/pokemon-list.module#PokemonListModule'
    },
    {
        path: 'pokemon',
        loadChildren:
            '../pokemon-details/pokemon-details.module#PokemonDetailsModule'
    }
];

@NgModule({
    imports: [RouterModule.forRoot(routes, { useHash: true })],
    exports: [RouterModule]
})
export class CoreRoutingModule {}
