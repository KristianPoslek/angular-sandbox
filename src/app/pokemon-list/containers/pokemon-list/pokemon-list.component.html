<div class="section">
    <h2 class="title is-1">Pokemon list</h2>

    <div class="columns is-multiline">
        <div class="column is-one-quarter" *ngFor="let pokemon of pokemonList">
            <div class="card">
                <div class="card-content">
                    <p class="title">
                        {{ pokemon.name }}
                    </p>
                </div>
                <footer class="card-footer">
                    <p class="card-footer-item">
                        <span>
                            <a [routerLink]="'/pokemon/' + pokemon.name"
                                >Check details</a
                            >
                        </span>
                    </p>
                </footer>
            </div>
        </div>
    </div>

    <pkmn-list-paging
        [previousUrl]="previousUrl"
        [nextUrl]="nextUrl"
        (paging)="fetchPokemonList($event)"
    ></pkmn-list-paging>
</div>
