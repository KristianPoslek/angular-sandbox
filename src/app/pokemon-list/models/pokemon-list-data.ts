import { Pokemon } from 'src/app/shared/models/pokemon';

export interface PokemonListData {
    lastRequestUrl: string;
    pokemonList: Array<Pokemon>;
    nextUrl: string;
    previousUrl: string;
}
