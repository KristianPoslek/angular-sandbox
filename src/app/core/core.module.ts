import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';
import { PokemonDetailsModule } from '../pokemon-details/pokemon-details.module';
import { PokemonListModule } from '../pokemon-list/pokemon-list.module';
import { SharedModule } from '../shared/shared.module';
import { AppComponent } from './containers/app/app.component';
import { CoreRoutingModule } from './core-routing.module';
import { metaReducers } from './state/meta.reducers';

@NgModule({
    declarations: [AppComponent],
    imports: [
        BrowserModule,
        CoreRoutingModule,
        SharedModule.forRoot(),
        PokemonListModule,
        PokemonDetailsModule,
        StoreModule.forRoot({}, { metaReducers }),
        EffectsModule.forRoot([])
    ]
})
export class CoreModule {}
