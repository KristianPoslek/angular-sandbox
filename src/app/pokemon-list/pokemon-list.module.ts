import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';
import { PokemonListComponent } from './containers/pokemon-list/pokemon-list.component';
import { PokemonListRoutingModule } from './pokemon-list.routing.module';
import { DataEffects } from './state/effects/data.effects';
import { reducers } from './state/reducers';
import { PokemonListStateService } from './state/services/pokemon-list.service';

@NgModule({
    declarations: [PokemonListComponent],
    imports: [
        BrowserModule,
        StoreModule.forFeature('pokemonList', reducers),
        EffectsModule.forFeature([DataEffects]),
        PokemonListRoutingModule
    ],
    providers: [PokemonListStateService]
})
export class PokemonListModule {}
