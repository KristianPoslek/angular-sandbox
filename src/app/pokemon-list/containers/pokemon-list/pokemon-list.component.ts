import { Component, OnDestroy, OnInit } from '@angular/core';
import { Subscription } from 'rxjs';
import { Pokemon } from 'src/app/shared/models/pokemon';
import { SharedStateService } from 'src/app/shared/state/services/shared-state.service';
import { PokemonListStateService } from '../../state/services/pokemon-list-state.service';

@Component({
    selector: 'pkmn-list',
    templateUrl: './pokemon-list.component.html'
})
export class PokemonListComponent implements OnInit, OnDestroy {
    subscriptions: Subscription = new Subscription();
    pokemonList: Pokemon[];
    previousUrl = '';
    nextUrl = '';

    constructor(
        private pokemonListStateService: PokemonListStateService,
        private sharedStateService: SharedStateService
    ) {}

    ngOnInit() {
        this.subscriptions.add(
            this.pokemonListStateService.getDataState().subscribe(data => {
                this.pokemonList = data.pokemonList;
                this.previousUrl = data.previousUrl;
                this.nextUrl = data.nextUrl;
            })
        );

        this.fetchPokemonList();
        this.sharedStateService.addLoading();
    }

    fetchPokemonList() {
        this.pokemonListStateService.fetchPokemonList();
    }

    ngOnDestroy() {
        this.subscriptions.unsubscribe();
    }
}
