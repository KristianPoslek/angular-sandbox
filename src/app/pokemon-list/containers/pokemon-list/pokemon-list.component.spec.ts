import { Injector } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { cold, Scheduler } from 'jest-marbles';
import { of } from 'rxjs';
import { SharedStateService } from 'src/app/shared/state/services/shared-state.service';
import { MockSharedStateService } from 'src/app/shared/state/services/shared-state.service.mock';
import { MockListPagingComponent } from '../../components/list-paging/list-paging.component.mock';
import { pokemonListDataMock } from '../../models/pokemon-list-data.mock';
import { PokemonListStateService } from '../../state/services/pokemon-list-state.service';
import { MockPokemonListStateService } from '../../state/services/pokemon-list-state.service.mock';
import { PokemonListComponent } from './pokemon-list.component';

describe('pokemonList.PokemonListComponent', () => {
    let testee: PokemonListComponent;
    let fixture: ComponentFixture<PokemonListComponent>;
    let fixtureInjector: Injector;
    let pokemonListStateServiceMock: PokemonListStateService;
    let sharedStateServiceMock: SharedStateService;

    beforeEach(async () => {
        TestBed.configureTestingModule({
            declarations: [PokemonListComponent, MockListPagingComponent],
            providers: [
                {
                    provide: PokemonListStateService,
                    useClass: MockPokemonListStateService
                },
                {
                    provide: SharedStateService,
                    useClass: MockSharedStateService
                }
            ],
            imports: [RouterTestingModule]
        });

        fixture = TestBed.createComponent(PokemonListComponent);
        testee = fixture.componentInstance;
        fixtureInjector = fixture.debugElement.injector;
        pokemonListStateServiceMock = fixtureInjector.get<
            PokemonListStateService
        >(PokemonListStateService);
        sharedStateServiceMock = fixtureInjector.get<SharedStateService>(
            SharedStateService
        );
    });

    describe('on component init', () => {
        beforeEach(() => {
            testee.fetchPokemonList = jest.fn();
            pokemonListStateServiceMock.getDataState = jest.fn(() => of());
        });

        it('should have a working snapshot', () => {
            fixture.detectChanges();
            expect(fixture).toMatchSnapshot();
        });

        it('should trigger pokemon list fetching and loading if there are no pokemon in state', () => {
            const pokemon$ = cold('--x', {
                x: { ...pokemonListDataMock, pokemonList: [] }
            });

            pokemonListStateServiceMock.getDataState = jest
                .fn()
                .mockReturnValue(pokemon$);

            testee.ngOnInit();
            Scheduler.get().flush();

            expect(testee.fetchPokemonList).toHaveBeenCalledTimes(1);
            expect(testee.fetchPokemonList).toHaveBeenCalledWith(
                pokemonListDataMock.lastRequestUrl
            );
            expect(sharedStateServiceMock.addLoading).toHaveBeenCalledTimes(1);
        });

        it('should not trigger pokemon list fetching and loading if there are pokemon in state', () => {
            const pokemon$ = cold('--x', {
                x: pokemonListDataMock
            });

            pokemonListStateServiceMock.getDataState = jest
                .fn()
                .mockReturnValue(pokemon$);

            testee.ngOnInit();
            Scheduler.get().flush();

            expect(testee.fetchPokemonList).not.toHaveBeenCalled();
            expect(sharedStateServiceMock.addLoading).not.toHaveBeenCalled();
        });

        it('should subscribe to pokemon list data changes', () => {
            const pokemon$ = cold('--x', {
                x: pokemonListDataMock
            });

            pokemonListStateServiceMock.getDataState = jest
                .fn()
                .mockReturnValue(pokemon$);

            testee.ngOnInit();
            Scheduler.get().flush();

            expect(testee.pokemonList).toEqual(pokemonListDataMock.pokemonList);
            expect(testee.previousUrl).toEqual(pokemonListDataMock.previousUrl);
            expect(testee.nextUrl).toEqual(pokemonListDataMock.nextUrl);
        });
    });

    it('should enable pokemon list fetching', () => {
        testee.fetchPokemonList();
        expect(pokemonListStateServiceMock.fetchPokemonList).toHaveBeenCalled();
    });

    it('should enable pokemon list fetching with URL', () => {
        testee.fetchPokemonList('foo');
        expect(
            pokemonListStateServiceMock.fetchPokemonList
        ).toHaveBeenCalledWith('foo');
    });

    it('should perform on destroy actions', () => {
        testee.subscriptions.unsubscribe = jest.fn();
        testee.ngOnDestroy();

        expect(testee.subscriptions.unsubscribe).toHaveBeenCalled();
    });
});
