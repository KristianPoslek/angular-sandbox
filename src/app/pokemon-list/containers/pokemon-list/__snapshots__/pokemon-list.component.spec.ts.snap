// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pokemonList.PokemonListComponent on component init should have a working snapshot 1`] = `
<pkmn-list
  fetchPokemonList={[Function Function]}
  nextUrl=""
  pokemonListStateService={[Function PokemonListStateServiceMock]}
  previousUrl=""
  sharedStateService={[Function SharedStateServiceMock]}
  subscriptions={[Function Subscription]}
>
  <div
    class="section"
  >
    <h2
      class="title is-1"
    >
      Pokemon list
    </h2>
    <div
      class="columns is-multiline"
    >
      
    </div>
    <pkmn-list-paging
      ng-reflect-next-url=""
      ng-reflect-previous-url=""
    >
      ./list-paging.component.html
    </pkmn-list-paging>
  </div>
</pkmn-list>
`;
